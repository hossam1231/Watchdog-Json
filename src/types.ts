import { JsonValue } from 'type-fest'


export enum DiffOperation {
  Added = 'A',
  Deleted = 'D',
  Moved = 'R',
  Modified = 'M'
}

export type MovedArrayItem = {
  diffType: DiffOperation.Moved
  index: number
  previousIndex: number
  value: JsonValue | undefined
}

export type DeletedArrayItem = {
  diffType: DiffOperation.Deleted
  previousIndex: number
  oldValue: JsonValue | undefined
}

export type ModifiedArrayItem = {
  diffType: DiffOperation.Modified
  index: number
  value: JsonValue | undefined
  oldValue: JsonValue | undefined
}

export type AddedArrayItem = {
  diffType: DiffOperation.Added
  index: number
  value: JsonValue | undefined
}

export type UnchangedArrayItem = {
  index: number
  value: JsonValue | undefined
}

export type ArrayDiffOperation =
  | UnchangedArrayItem
  | MovedArrayItem
  | DeletedArrayItem
  | ModifiedArrayItem
  | AddedArrayItem

export type UnchangedObjectItem = {
  key: string
  value: JsonValue | undefined
}

export type AddedObjectItem = {
  diffType: DiffOperation.Added
  key: string
  value: JsonValue | undefined
}

export type ModifiedObjectItem = {
  diffType: DiffOperation.Modified
  key: string
  value: JsonValue | undefined
  oldValue: JsonValue | undefined
}

export type DeletedObjectItem = {
  diffType: DiffOperation.Deleted
  key: string
  oldValue: JsonValue | undefined
}

export type ObjectDiffOperation = UnchangedObjectItem | AddedObjectItem | ModifiedObjectItem | DeletedObjectItem

export type ObjectAnalysis = {
  type: 'object'
  decorations: ['{', '}']
  items: ObjectDiffOperation[]
}

export type ArrayAnalysis = {
  type: 'array'
  decorations: ['[', ']']
  items: ArrayDiffOperation[]
}

export type ValueAnalysis = ArrayAnalysis | ObjectAnalysis
