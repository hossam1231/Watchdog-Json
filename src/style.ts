import shader from 'shader'
import styled, { css } from 'styled-components'

import { DiffOperation } from './types'


const ArrowRightIcon = `
  <svg
    width="200mm"
    height="200mm"
    viewBox="0 0 200 200"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:svg="http://www.w3.org/2000/svg">
    <path
        style="fill:none;stroke:#52bd94;stroke-width:25;stroke-linecap:round;stroke-linejoin:round;stroke-dasharray:none"
        d="m 75.60726,44.41028 44.12533,44.125337 c 6.35127,6.351268 6.35127,16.577493 1e-5,22.928753 l -44.125343,44.12535"/>
  </svg>
`.replaceAll(/#/g, '%23')

const createShades = (color: string, shades: number = 3) =>
  Array.from({ length: shades }, (_, i) => shader(color, i === 0 ? 0 : i / shades))

const generateBands = (color1: string, color2: string) => css`
  background: repeating-linear-gradient(-45deg, ${color1}, ${color1} 4px, ${color2} 4px, ${color2} 8px);
`
const colors: Partial<Record<DiffOperation, string[]>> = {
  [DiffOperation.Added]: createShades('#0ec080'),
  [DiffOperation.Deleted]: createShades('#ff3232'),
  [DiffOperation.Modified]: createShades('#0e56c0')
}

export const ObjectKey = styled.span`
  font-family: monospace;
  color: #004b7c;
`

export const ArrayIndex = styled(ObjectKey)<{ moved?: boolean }>`
  ${({ moved }) =>
    moved &&
    css`
      background-color: #d3e1ff;
    `}

  &::before,
  &::after {
    color: grey;
  }

  &::before {
    content: '[';
  }

  &::after {
    content: ']';
  }
`

export const PrevArrayIndex = styled(ArrayIndex)`
  background-color: #ff9f9f;
  color: #ab0101;
  text-decoration: line-through;

  &::before,
  &::after {
    color: grey;
  }
`

export const ColonSeparator = styled.span`
  &::after {
    content: ':';
    font-family: monospace;
    margin-right: 1ch;
    color: #343434;
  }
`

export const PrimitiveValue = styled.span<{ rawValue: unknown; diffType?: DiffOperation }>`
  position: relative;
  font-family: monospace;
  display: inline-block;
  margin: 1px;

  ${({ diffType }) =>
    diffType &&
    css`
      ${generateBands(shader(colors[diffType]![0], 0.7), shader(colors[diffType]![2], 0.7))}
      background-attachment: scroll;

      ${diffType === DiffOperation.Deleted &&
      css`
        text-decoration: line-through;
      `}
    `}

  ${({ rawValue }) =>
    typeof rawValue === 'string' &&
    css`
      color: #8b3f00;
      &::before,
      &::after {
        content: '"';
        color: #c9986a;
      }
    `}

  ${({ rawValue }) =>
    typeof rawValue === 'number' &&
    css`
      color: #015dd2;
    `}
  
  ${({ rawValue }) =>
    typeof rawValue === 'boolean' &&
    css`
      color: #7c01b7;
    `}
  
  ${({ rawValue }) =>
    (typeof rawValue === 'undefined' || rawValue === null) &&
    css`
      color: grey;
    `}
`

export const ToggleObjectButton = styled.button`
  background: none;
  border: none;
  outline: none;
  position: absolute;
  left: -1em;
  width: 1em;
  height: 1em;
  padding: 0;
`

export const HighlightDecoration = styled.div`
  width: 5px;
  background: #428cdc;
  display: block;
  position: absolute;
  left: -19px;
  top: 0;
  bottom: 0;
`

export const DefaultToggleIcon = styled.div<{ collapsed: boolean }>`
  height: 100%;
  width: 100%;
  background: center / 100% no-repeat url('data:image/svg+xml;utf8,${ArrowRightIcon}');

  ${({ collapsed }) =>
    !collapsed &&
    css`
      transform: rotate(45deg);
    `}
`

export const LabelDecorations = styled.div`
  display: inline-flex;
  align-items: center;
  position: relative;
`

export const CollapsedIndicator = styled.span`
  position: relative;

  &::before {
    content: '...';
    color: #838383;
  }
`

export const BracketDecoration = styled.span`
  display: inline-block;
  position: relative;
  font-family: monospace;
  color: #969696;
  font-style: italic;

  /* the opening bracket should have just margin-top, and closing bracket should have just margin-bottom */
  margin-top: 1px;
  & ~ & {
    margin-top: 0;
    margin-bottom: 1px;
  }
`

export const ItemList = styled.ol`
  padding: 0 0 0 20px;
  margin: 0;
  list-style: none;
  position: relative;

  &::before {
    content: '';
    display: block;
    position: absolute;
    width: 1px;
    inset: 0.5ch 0 0.5ch 0.3ch;
    border-right: 1px solid #dfdfdf;
  }
`

export const DiffOverlay = styled.div<{ dimmingFactor: number; diffType?: DiffOperation }>`
  position: relative;
  z-index: 2;
  display: inline-block;

  ${({ diffType, dimmingFactor }) =>
    typeof diffType !== 'undefined' &&
    diffType in colors &&
    css`
      &::before,
      &::after {
        content: '';
        display: block;
        position: absolute;
        inset: 0px -4px 0px -2px;
        border: 1px solid #fff;
        border-radius: 4px;
        transition: 0.2s opacity ease-in;
      }

      /* this "backdrop" prevents mixing of colors across mutiple diff highlight layers */
      /* this also prevents mixing of the layer color with the background color of container */
      &::before {
        z-index: -1;
        background-color: #fff;
        opacity: ${dimmingFactor > 0.5 ? 1 : 0};
      }

      &::after {
        z-index: -1;
        opacity: ${0.3 * dimmingFactor};
        border-color: ${colors[diffType]![0]};
        background: repeating-linear-gradient(
          -45deg,
          ${colors[diffType]![0]},
          ${colors[diffType]![0]} 4px,
          ${colors[diffType]![2]} 4px,
          ${colors[diffType]![2]} 8px
        );
        background-attachment: scroll;
      }
    `}
`

export const Item = styled.li`
  margin: 2px 0;
  padding-right: 4px;
`
