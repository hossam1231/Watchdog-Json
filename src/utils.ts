import { JsonArray, JsonObject, JsonValue } from 'type-fest'
import isObject from 'is-object'
import mapObject from 'map-obj'

import {
  AddedArrayItem,
  AddedObjectItem,
  ArrayDiffOperation,
  ValueAnalysis,
  DeletedArrayItem,
  DeletedObjectItem,
  DiffOperation,
  ModifiedArrayItem,
  ModifiedObjectItem,
  MovedArrayItem,
  ObjectDiffOperation,
  UnchangedObjectItem
} from './types'


const isRecord = (v: unknown): v is Record<string, unknown> => isObject(v) && !Array.isArray(v)
const hash = (v: JsonValue) => JSON.stringify(v)
const difference = <T>(a: T[], b: T[]) => a.filter((k) => !b.includes(k))
const intersection = <T>(a: T[], b: T[]) => a.filter((hash) => b.includes(hash))
const isSameValueType = <T>(v1: T, v2: unknown): v2 is T =>
  (isRecord(v1) && isRecord(v2)) || (Array.isArray(v1) && Array.isArray(v2)) || typeof v1 === typeof v2

const compareArrays = <L extends JsonArray, R extends JsonArray>(left: L, right: R): ArrayDiffOperation[] | void => {
  if (Array.isArray(left) && Array.isArray(right)) {
    const hashedLeft = left.map(hash)
    const hashedRight = right.map(hash)
    const additions = difference(hashedRight, hashedLeft).map((hash) => hashedRight.indexOf(hash))
    const deletions = difference(hashedLeft, hashedRight).map((hash) => hashedLeft.indexOf(hash))
    const moves = intersection(hashedRight, hashedLeft)
      .map((hash) => [hashedLeft.indexOf(hash), hashedRight.indexOf(hash)])
      .filter(([indexInLeft, indexInRight]) => indexInLeft !== indexInRight)
    const modifications = deletions.reduce(
      (acc, index) =>
        additions.includes(index) && isSameValueType(left[index], right[index])
          ? [
              ...acc,
              {
                diffType: DiffOperation.Modified,
                index: index,
                oldValue: left[index],
                value: right[index]
              }
            ]
          : acc,
      [] as ModifiedArrayItem[]
    )

    const changes: ArrayDiffOperation[] = [
      ...modifications,
      ...deletions
        .filter((deletionIndex) => !modifications.some(({ index }) => index === deletionIndex))
        .map<DeletedArrayItem>((deletionIndex) => ({
          diffType: DiffOperation.Deleted,
          previousIndex: deletionIndex,
          oldValue: left[deletionIndex]
        })),
      ...additions
        .filter((additionIndex) => !modifications.some(({ index }) => index === additionIndex))
        .map<AddedArrayItem>((additionIndex) => ({
          diffType: DiffOperation.Added,
          index: additionIndex,
          value: right[additionIndex]
        })),
      ...moves.map<MovedArrayItem>(([indexInLeft, indexInRight]) => ({
        diffType: DiffOperation.Moved,
        index: indexInRight,
        previousIndex: indexInLeft,
        value: right[indexInRight]
      }))
    ]

    const base = right.map<ArrayDiffOperation>((value, index) => ({ index, value }))

    // transplant changes into base
    changes.forEach((change) => {
      if ('index' in change) {
        return base.splice(
          base.findIndex((item) => 'index' in item && item.index === change.index),
          1,
          change
        )
      }

      // Delete operation
      if ('previousIndex' in change) {
        return base.splice(
          base.findIndex((item) => 'index' in item && item.index === change.previousIndex),
          0,
          change
        )
      }
    })

    return base
  }

  // when only one of left or right is an array, there's no comparison
  const subject = (Array.isArray(right) && right) || (Array.isArray(left) && left)
  return !Array.isArray(subject)
    ? undefined
    : subject.map((v, i) => ({
        index: i,
        value: v
      }))
}

const compareObjects = <L extends JsonValue, R extends JsonValue>(left: L, right: R): ObjectDiffOperation[] | void => {
  if (isRecord(left) && isRecord(right)) {
    const hashedLeft = mapObject(left as Record<string, JsonValue>, (key, value) => [key, hash(value)])
    const hashedRight = mapObject(right as Record<string, JsonValue>, (key, value) => [key, hash(value)])
    const leftKeys = Object.keys(hashedLeft)
    const rightKeys = Object.keys(hashedRight)
    const addedKeys = difference(rightKeys, leftKeys)
    const deletedKeys = difference(leftKeys, rightKeys)
    const intersectingKeys = intersection(rightKeys, leftKeys)
    const modifiedKeys = intersectingKeys.filter((k) => hashedLeft[k] !== hashedRight[k])
    const unchangedKeys = difference(intersectingKeys, modifiedKeys)

    const getKeyIndex = (k: string) => {
      const idx = rightKeys.indexOf(k)
      return idx > -1 ? idx : leftKeys.indexOf(k)
    }

    return [
      ...addedKeys.map<AddedObjectItem>((key) => ({
        diffType: DiffOperation.Added,
        key,
        value: right[key]
      })),
      ...deletedKeys.map<DeletedObjectItem>((key) => ({
        diffType: DiffOperation.Deleted,
        key,
        oldValue: left[key]
      })),
      ...modifiedKeys.flatMap<ModifiedObjectItem | DeletedObjectItem | AddedObjectItem>((key) =>
        isSameValueType(left[key], right[key])
          ? {
              diffType: DiffOperation.Modified,
              key,
              value: right[key],
              oldValue: left[key]
            }
          : [
              {
                diffType: DiffOperation.Deleted,
                key,
                oldValue: left[key]
              },
              {
                diffType: DiffOperation.Added,
                key,
                value: right[key]
              }
            ]
      ),
      ...unchangedKeys.map<UnchangedObjectItem>((key) => ({
        key,
        value: right[key]
      }))
    ].sort((a, b) => getKeyIndex(a.key) - getKeyIndex(b.key))
  }

  // when only one of left or right is an Object, there's no comparison
  const subject = (isRecord(right) && right) || (isRecord(left) && left)
  return !isRecord(subject)
    ? undefined
    : Object.entries(subject).map<UnchangedObjectItem>(([k, v]) => ({
        key: k,
        value: v as JsonValue
      }))
}

export const compare = <
  L extends JsonValue | Record<string, unknown> | void,
  R extends JsonValue | Record<string, unknown> | void
>(
  left: L,
  right: R
): ValueAnalysis | void => {
  if (Array.isArray(right) || Array.isArray(left)) {
    return {
      type: 'array',
      decorations: ['[', ']'],
      items: compareArrays(left as JsonArray, right as JsonArray)!
    }
  }

  if (isRecord(right) || isRecord(left)) {
    return {
      type: 'object',
      decorations: ['{', '}'],
      items: compareObjects(left as JsonObject, right as JsonObject)!
    }
  }
}
