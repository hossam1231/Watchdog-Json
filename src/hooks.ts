import { useMemo } from 'react'
import { JsonValue } from 'type-fest'

import { ValueAnalysis } from './types'
import { compare } from './utils'


type Value = JsonValue | Record<string, unknown> | undefined

export const useAugmentedValue = (left: Value, right: Value): ValueAnalysis | void =>
  useMemo(() => compare(left, right), [left, right])
