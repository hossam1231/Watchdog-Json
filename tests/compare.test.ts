import { compare } from '../src/utils'
import {
  ArrayDiffOperation,
  DiffOperation,
  ModifiedArrayItem,
  ModifiedObjectItem,
  ObjectDiffOperation
} from '../src/types'


const originalValue = Object.freeze({
  /**
   * This will be a very basic test for primitives being modified
   */
  thePrimitives: {
    aString: 'jacob',
    aLongString: 'brown fox jumped over the fence, she fell but she is alive and well, and no regrets',
    aNumericValue: 7329821,
    aBoolean: true
  },
  /**
   * This should test for elements being modified (to the same value type, and other types), removals,
   * elements being moved around without being changed, additions
   */
  anArray: [
    829502,
    889252,
    'armenia',
    'this value switches positions',
    { thisObjectSwitchesPositions: true },
    // a value is expected to be added somewhere here
    false
  ],
  /**
   * This should test for inner changes to complex values within array
   *
   * This assertion absolutely requires that elements preserve their position AND the type
   * in the updated array to be eligible for direct comparison
   */
  anArrayWithInnerChanges: [
    'stable value as experiment control',
    { productId: 48925, description: 'a product' },
    'another stable value as experiment control',
    [1, 2, 3]
  ],
  /**
   * This should test for object having inner changes - keys being modified, removed and additions
   */
  anObject: {
    thisKeyWillBeModified: {
      name: 'tuco salamanca',
      age: 42
    },
    thisKeyWillBeRemoved: {
      name: 'sussy baca',
      age: 21
    }

    // a key is expected be added here
  }
})

const changedValue = Object.freeze({
  thePrimitives: {
    aString: 'ryan',
    aLongString: 'brown fox jumped over the fence, there are regrets',
    aNumericValue: 121,
    aBoolean: false
  },
  anArray: [
    174211,
    { thisObjectSwitchesPositions: true },
    'japan',
    true,
    'this value has been added',
    'this value switches positions'
  ],
  anArrayWithInnerChanges: [
    'stable value as experiment control',
    { productId: 73211, description: 'a product', catalogId: 810985 },
    'another stable value as experiment control',
    [3, 4, 5]
  ],
  anObject: {
    thisKeyWillBeModified: {
      name: 'lalo salamanca',
      age: 39
    },
    aNewlyAddedKey: {
      name: 'saul goodman',
      age: 35
    }
  }
})

const allDiffOperations = Object.values(DiffOperation).filter((v) => typeof v === 'number') as DiffOperation[]

describe('JSON Compare', () => {
  it('performs comparison of basic primitives', () => {
    const depth0Result = compare(originalValue, changedValue)!

    const item = depth0Result.items[0] as ModifiedObjectItem

    expect(item).toMatchObject(
      expect.objectContaining({
        diffType: DiffOperation.Modified,
        key: 'thePrimitives'
      })
    )

    const subject = compare(item.oldValue, item.value)

    expect(subject).toMatchSnapshot('primitives_diff_result')
  })

  it('performs comparison of an array', () => {
    const depth0Result = compare(originalValue, changedValue)!

    const item = depth0Result.items[1] as ModifiedObjectItem

    expect(item).toMatchObject(
      expect.objectContaining({
        diffType: DiffOperation.Modified,
        key: 'anArray'
      })
    )

    const subject = compare(item.oldValue, item.value)!

    // this assertion makes sure that the test data has been set up to test all types of diff operations
    expect(
      allDiffOperations.every((diffType) =>
        (subject.items as ArrayDiffOperation[]).some((item) => 'diffType' in item && item.diffType === diffType)
      )
    ).toBe(true)

    expect(subject).toMatchSnapshot('array_diff_result')
  })

  it('performs inner comparison of an array', () => {
    const depth0Result = compare(originalValue, changedValue)!

    const item = depth0Result.items[2] as ModifiedObjectItem

    expect(item).toMatchObject(
      expect.objectContaining({
        diffType: DiffOperation.Modified,
        key: 'anArrayWithInnerChanges'
      })
    )

    const subject = compare(item.oldValue, item.value)!

    const innerItem1 = subject.items[1] as ModifiedObjectItem
    const innerItem2 = subject.items[3] as ModifiedArrayItem

    // this assertion proves that as long the subjects don't move around, differ can do a direct compare
    expect(innerItem1.diffType).toBe(DiffOperation.Modified)
    expect(innerItem2.diffType).toBe(DiffOperation.Modified)

    const innerItem1Subject = compare(innerItem1.oldValue, innerItem1.value)!

    expect(innerItem1Subject).toMatchSnapshot('array_inner_object_diff_result')

    const innerItem2Subject = compare(innerItem2.oldValue, innerItem2.value)

    expect(innerItem2Subject).toMatchSnapshot('array_inner_array_diff_result')
  })

  it('performs comparison of an object', () => {
    const depth0Result = compare(originalValue, changedValue)!

    const item = depth0Result.items[3] as ModifiedObjectItem

    expect(item).toMatchObject(
      expect.objectContaining({
        diffType: DiffOperation.Modified,
        key: 'anObject'
      })
    )

    const subject = compare(item.oldValue, item.value)!

    // this assertion makes sure that the test data has been set up to test all types of diff operations
    expect(
      allDiffOperations.every(
        (diffType) =>
          (subject.items as ObjectDiffOperation[]).some((item) => 'diffType' in item && item.diffType === diffType) ||
          diffType === DiffOperation.Moved // there's no "moved" diff operation in objects
      )
    ).toBe(true)

    expect(subject).toMatchSnapshot('object_diff_result')
  })
})
