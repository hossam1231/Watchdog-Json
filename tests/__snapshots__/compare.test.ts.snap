// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Compare performs comparison of an array: array_diff_result 1`] = `
Object {
  "decorations": Array [
    "[",
    "]",
  ],
  "items": Array [
    Object {
      "diffType": "M",
      "index": 0,
      "oldValue": 829502,
      "value": 174211,
    },
    Object {
      "diffType": "D",
      "oldValue": 889252,
      "previousIndex": 1,
    },
    Object {
      "diffType": "R",
      "index": 1,
      "previousIndex": 4,
      "value": Object {
        "thisObjectSwitchesPositions": true,
      },
    },
    Object {
      "diffType": "M",
      "index": 2,
      "oldValue": "armenia",
      "value": "japan",
    },
    Object {
      "diffType": "A",
      "index": 3,
      "value": true,
    },
    Object {
      "diffType": "A",
      "index": 4,
      "value": "this value has been added",
    },
    Object {
      "diffType": "D",
      "oldValue": false,
      "previousIndex": 5,
    },
    Object {
      "diffType": "R",
      "index": 5,
      "previousIndex": 3,
      "value": "this value switches positions",
    },
  ],
  "type": "array",
}
`;

exports[`JSON Compare performs comparison of an object: object_diff_result 1`] = `
Object {
  "decorations": Array [
    "{",
    "}",
  ],
  "items": Array [
    Object {
      "diffType": "M",
      "key": "thisKeyWillBeModified",
      "oldValue": Object {
        "age": 42,
        "name": "tuco salamanca",
      },
      "value": Object {
        "age": 39,
        "name": "lalo salamanca",
      },
    },
    Object {
      "diffType": "A",
      "key": "aNewlyAddedKey",
      "value": Object {
        "age": 35,
        "name": "saul goodman",
      },
    },
    Object {
      "diffType": "D",
      "key": "thisKeyWillBeRemoved",
      "oldValue": Object {
        "age": 21,
        "name": "sussy baca",
      },
    },
  ],
  "type": "object",
}
`;

exports[`JSON Compare performs comparison of basic primitives: primitives_diff_result 1`] = `
Object {
  "decorations": Array [
    "{",
    "}",
  ],
  "items": Array [
    Object {
      "diffType": "M",
      "key": "aString",
      "oldValue": "jacob",
      "value": "ryan",
    },
    Object {
      "diffType": "M",
      "key": "aLongString",
      "oldValue": "brown fox jumped over the fence, she fell but she is alive and well, and no regrets",
      "value": "brown fox jumped over the fence, there are regrets",
    },
    Object {
      "diffType": "M",
      "key": "aNumericValue",
      "oldValue": 7329821,
      "value": 121,
    },
    Object {
      "diffType": "M",
      "key": "aBoolean",
      "oldValue": true,
      "value": false,
    },
  ],
  "type": "object",
}
`;

exports[`JSON Compare performs inner comparison of an array: array_inner_array_diff_result 1`] = `
Object {
  "decorations": Array [
    "[",
    "]",
  ],
  "items": Array [
    Object {
      "diffType": "D",
      "oldValue": 1,
      "previousIndex": 0,
    },
    Object {
      "diffType": "R",
      "index": 0,
      "previousIndex": 2,
      "value": 3,
    },
    Object {
      "diffType": "M",
      "index": 1,
      "oldValue": 2,
      "value": 4,
    },
    Object {
      "diffType": "A",
      "index": 2,
      "value": 5,
    },
  ],
  "type": "array",
}
`;

exports[`JSON Compare performs inner comparison of an array: array_inner_object_diff_result 1`] = `
Object {
  "decorations": Array [
    "{",
    "}",
  ],
  "items": Array [
    Object {
      "diffType": "M",
      "key": "productId",
      "oldValue": 48925,
      "value": 73211,
    },
    Object {
      "key": "description",
      "value": "a product",
    },
    Object {
      "diffType": "A",
      "key": "catalogId",
      "value": 810985,
    },
  ],
  "type": "object",
}
`;
